necsim/plugins/tskit/src/tree/table.rs

```diff
@@ -0,0 +1,183 @@
+                self.store_lineage(&child, time, Some((parent_individual, parent_node)))
+            {
+                self.store_children_of_parent(&child, child_individual, child_node);
+            }
+        } else {
+            // The parent has not been inserted yet
+            //  -> postpone insertion and remember the child
+            self.children.entry(parent).or_default().push((child, time));
+        }
+    }
+
+    pub(super) fn store_provenance(&mut self) -> Result<(), String> {
+        // Capture and record the provenance information inside the table
+        let provenance =
+            crate::provenance::TskitProvenance::try_new().map_err(|err| err.to_string())?;
+        let provenance_json = serde_json::to_string(&provenance).map_err(|err| err.to_string())?;
+
+        self.table
+            .add_provenance(&provenance_json)
+            .map_err(|err| err.to_string())
+            .map(|_| ())
+    }
+
+    pub(super) fn output_tree_sequence(mut self) {
+        self.table.full_sort(TableSortOptions::NONE).unwrap();
+
+        // Output the tree sequence to the specified `output` file
+        self.table
+            .tree_sequence(TreeSequenceFlags::BUILD_INDEXES)
+            .unwrap()
+            .dump(&self.output, TableOutputOptions::NONE)
+            .unwrap();
+    }
+}
+
+impl TskitTreeReporter {
+    /// Store a lineage as a `tskit` individual and birth node, optionally with
+    /// a parent relationship
+    fn store_lineage(
+        &mut self,
+        reference: &GlobalLineageReference,
+        time: NonNegativeF64,
+        parent: Option<(IndividualId, NodeId)>,
+    ) -> Option<(IndividualId, NodeId)> {
+        let origin = self.origins.remove(reference)?;
+        let location = [
+            f64::from(origin.location().x()),
+            f64::from(origin.location().y()),
+            f64::from(origin.index()),
+        ];
+        let metadata = GlobalLineageMetadata::new(reference);
+        let parents = if let Some((parent_individual, _parent_node)) = &parent {
+            std::slice::from_ref(parent_individual)
+        } else {
+            &[]
+        };
+
+        // Insert the lineage as an individual
+        let individual_id = self
+            .table
+            .add_individual_with_metadata(TSK_FLAGS_EMPTY, &location, parents, metadata)
+            .unwrap();
+
+        // Create corresponding node
+        let node_id = self
+            .table
+            .add_node_with_metadata(
+                tskit::TSK_NODE_IS_SAMPLE,
+                time.get(),
+                tskit::PopulationId::NULL,
+                individual_id,
+                metadata,
+            )
+            .unwrap();
+
+        if let Some((_parent_individual, parent_node)) = parent {
+            // Add the parent-child relation between the nodes
+            self.table
+                .add_edge(TSK_SEQUENCE_MIN, TSK_SEQUENCE_MAX, parent_node, node_id)
```

Comment:
if I understand right, your simulations have no underlying genetics?  so here, parent and node are references to "individuals" and the positions are dummy values to make the data structures happy?

Reply:
Yes, there are no underlying genetics in my simulation. I use nodes to refer to lineages, which in my simulation are represented by the present-time individuals they originate from, and which span many generations (and thereby individuals) until the lineage either coalesces or speciates (slightly simplified as lineages can also overlap after coalescence instead of one lineage just 'stopping')

Reply:
So I think this PR is definitely merge-eable.  One thing I didn't check/notice is if there's an API to mark nodes as "samples" in a flexible way, facilitating downstream analyses?

Reply:
Yes, in `necsim/plugins/tskit/src/tree/table.rs` line 137 I set every node to be a sample